xray/src/client.rs

```diff
@@ -0,0 +1,222 @@
+            SocketAddr::V4(addr) => addr,
+            SocketAddr::V6(_) => panic!("IPv6 not supported"),
+        };
+        let mut ret = if self.buf[0] == 2 {
+            RecvType::Handshake
+        } else if self.buf[0] == 4 {
+            RecvType::Data { from, length: 0 }
+        } else {
+            panic!("Unexpected packet type {}", self.buf[0])
+        };
+        let mut decap_buf = vec![0; 1024];
+        loop {
+            let tr = self.tunn.as_mut().unwrap().decapsulate(
+                None,
+                &self.buf[0..bytes_read],
+                &mut decap_buf,
+            );
+            match tr {
+                TunnResult::Done => break,
+                TunnResult::WriteToNetwork(p) => {
+                    self.sock.send_to(p, from).await.unwrap();
+                    bytes_read = 0;
+                }
+                TunnResult::WriteToTunnelV4(p, _) => {
+                    let (pfrom, payload_start, payload_end) = Self::parse_udp_packet(p);
+                    assert!(buf.len() >= payload_end - payload_start);
+                    buf.iter_mut()
+                        .zip(p[payload_start..payload_end].iter())
+                        .for_each(|(b, p)| *b = *p);
+                    ret = RecvType::Data {
+                        from: pfrom,
+                        length: payload_end - payload_start,
+                    };
+                    break;
+                }
+                unexpected => panic!("Unexpected TunnResult: {unexpected:?}"),
+            }
+        }
+        ret
+    }
+
+    async fn recv_plaintext(&mut self, buf: &mut [u8]) -> RecvType {
+        let (bytes_read, from) = self.sock.recv_from(buf).await.unwrap();
+        let from = match from {
+            SocketAddr::V4(addr) => addr,
+            SocketAddr::V6(_) => panic!("IPv6 not supported"),
+        };
+        RecvType::Data {
+            length: bytes_read,
+            from,
+        }
+    }
+
+    fn make_udp_packet(src: SocketAddrV4, dst: SocketAddrV4, payload: &[u8]) -> Vec<u8> {
+        let len = 28 + payload.len(); // IP header (20 bytes) + UDP heder (8 bytes) + payload length
+        let mut udp_packet = vec![0; len];
+        {
+            let mut ipv4 = MutableIpv4Packet::new(&mut udp_packet).unwrap();
+            ipv4.set_version(4);
+            ipv4.set_header_length(5);
+            ipv4.set_total_length(len as u16);
+            ipv4.set_ttl(20);
+            ipv4.set_next_level_protocol(IpNextHeaderProtocols::Udp);
+            ipv4.set_source(*src.ip());
+            ipv4.set_destination(*dst.ip());
+            ipv4.set_checksum(ipv4::checksum(&ipv4.to_immutable()));
+        }
+        {
+            let mut udp = MutableUdpPacket::new(&mut udp_packet[20..]).unwrap();
+            udp.set_source(src.port());
+            udp.set_destination(dst.port());
+            udp.set_length(8 + payload.len() as u16);
+            udp.set_payload(payload);
+            // TODO: Calculate checksum
```

Comment:
Given the new review guidelines, this TODO should contain your name or the ticket number:
// TODO(mathias):  or // TODO(LLT-9876):

Reply:
Fixed

Reply:
I just implemented the checksum so the `TODO` was removed